libass/ass_render.c

```diff
@@ -1801,8 +1851,28 @@ wrap_lines_measure(ASS_Renderer *render_priv)
 static void
 wrap_lines_smart(ASS_Renderer *render_priv, double max_text_width)
 {
-    wrap_lines_naive(render_priv, max_text_width);
-    wrap_lines_rebalance(render_priv, max_text_width);
+    char *unibrks = NULL;
+
+#ifdef CONFIG_UNIBREAK
+    if (render_priv->track->parser_priv->wrap_unicode) {
+        if (render_priv->scratch_buf_len < render_priv->text_info.length) {
```

Comment:
We have several allocations per glyph, so one per event doesn't really matter much. Better to keep code more simpler.

Reply:
Those other allocations aren't as short-lived though, i think? *(otherwise, if they also only live for a single function only calling its own internal, non-allocating sub-routines, they could also use the same scratch buffer introduced here instead)*

Reply:
I think persisting a buffer is fine, and we do it in other places such as the shaper. It might use a more descriptive name though, and it probably should live inside `render_priv->text_info`.

But I suppose it’s also true that it doesn’t make much of a difference.

Reply:
We do have short-lived buffers in blur (both be_blur and Gaussian). But I've looked into libunibreak and discovered that `set_linebreaks()` is a simple wrapper around `lb_init_break_context()` and `lb_process_next_char()`. Better to use these functions directly and store result somewhere in `GlyphInfo`, maybe even do that inside of `wrap_lines_naive()` loop to improve cache locality.

Reply:
While looking at the `lb_process_…` function, which doesn't exist in 1.1, I noticed something else: while `set_linebreaks`' signature remained unchanged from libunibreak 1.0 to the newest release 4.3, on unreleased master a [recent-ish commit](https://github.com/adah1972/libunibreak/commit/a6bcee2daf6fb884edd1ff78ce58521ab31f9826) changed it. [Another recent-ish commit](https://github.com/adah1972/libunibreak/commit/3df7eccb74474fb3cdec1da5365a35fc6f3c5da0) suggests, that while `linebreakdef.h` is distributed, it is actually considered to be a *private* header.
Perhaps it is better to not use anything from `linebreakdef.h`. This means we can't provide a `next_glyph` function. To not have to copy all glyphs into a continuos char array, calling `set_linebreaks_utf32` for each glyph individually seems like the safest option.

Reply:
On the other hand, that also seems like a bad idea, since this robs the glyphs of context and libunibreak already discourages the use of `is_line_breakable` which at least has two glyphs to work with.

Reply:
It's not good that it's not clearly documented which headers/functions are considered public.
Maybe we should create feature request for making public `lb_init_break_context()` and `lb_process_next_char()` as the most generic interface? Also, why `init` treats the first character differently from the rest? Is it just a unicode quirk?

Reply:
> To not have to copy all glyphs into a continuos char array, calling `set_linebreaks_utf32` for each glyph individually seems like the safest option.

FWIW we already do copy characters into a continuous char array for shaping, in `ass_shaper_shape`.

Reply:
Re: buffer name and location
I initially chose the generic name to allow the buffer to be reused for at other places. However, the mentioned buffers in be/blur appear to not have easy acces to the `render_priv`, so I renamed and moved it now.